# Quang Vu


Theorem 4.8 (Korselt criterion for Carmichael numbers). An integer n is a
Carmichael number if and only if n is

* positive,
* composite,
* square-free, and
* for each prime p dividing n, p − 1 also divides n − 1.


**Exploration 4.46:** Using Theorem 4.8, implement a function which will determine
whether a given value n is a Carmichael number. Use this function to find all the
Carmichael numbers up to 1,000,000 (or more).
How many Carmichael numbers are there less than 25,000,000?

Note: An effective implementation of Theorem 4.8 will utilize the prime factorization of the value n. Many programming languages such as Mathematica and Python have built-in functions which will do this in an efficient manner.

This function first checks if n is less than or equal to 2 or if it is even or has a square factor, in which case it cannot be a Carmichael number. Then, for each prime factor p of n, it checks if p-1 divides n-1, as required by the Korselt criterion.

In [ ]:
from sympy import factorint, isprime
import math

In [ ]:
def carmichaelNumbers(n):
    if n <= 2 or isprime(n): #positive values and prime
        return False
    factors = factorint(n)
    if any(exp > 1 for exp in factors.values()): #Checing if a number is composite or not
        return False
    for p in factors:
        if (n-1) % (p-1) != 0: #each prime p dividing n, p − 1 also divides n − 1.
            return False
    return True

In [ ]:
carmichaels = []
for n in range(3, 1000):
    if carmichaelNumbers(n):
        carmichaels.append(n)
print(carmichaels)


[561]


Using The On-Line Encyclopedia of Integer Sequences, I know that the first number for Carmichael is 561. As a rreuslt, I think the code is working. This code utilize some of python math library like checking prime number and prime factoriztion. Below,, I claculate Carmichael numbers up to 1,000,000.

In [ ]:
carmichaels1 = []
for n in range(3, 1000001):
    if carmichaelNumbers(n):
        carmichaels1.append(n)
print(carmichaels1)
print(len(carmichaels1))

[561, 1105, 1729, 2465, 2821, 6601, 8911, 10585, 15841, 29341, 41041, 46657, 52633, 62745, 63973, 75361, 101101, 115921, 126217, 162401, 172081, 188461, 252601, 278545, 294409, 314821, 334153, 340561, 399001, 410041, 449065, 488881, 512461, 530881, 552721, 656601, 658801, 670033, 748657, 825265, 838201, 852841, 997633]
43


The next code cell I want to see how many Carmichael numbers less than 25,000,000.

In [ ]:
carmichaels2 = []
for n in range(1, 25000001):
    if carmichaelNumbers(n):
        carmichaels2.append(n)
#print(carmichaels2)
print(len(carmichaels2))

145


I looks like there are 154 Carmicheal number less than 25,000,000

We suppose to check if there are exactly 2,163 Carmichael numbers less than 25,000,0000. Due to computational restriction, we can not compute to 25 billion numbers.